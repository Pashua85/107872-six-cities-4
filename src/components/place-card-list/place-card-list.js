import React from 'react';
import PropTypes from 'prop-types';
import CitiesPlaceCard from '../cities-place-card/cities-place-card';
import NearPlaceCard from '../near-place-card/near-place-card';

class PlaceCardList extends React.PureComponent {
  constructor(props) {
    super(props);
    this.handleCardHover = this.handleCardHover.bind(this);
    this.handleCardUnhover = this.handleCardUnhover.bind(this);

    this.state = {
      activeCard: null
    };
  }

  handleCardHover(place) {
    this.setState({
      activeCard: place
    });
  }

  handleCardUnhover() {
    this.setState({
      activeCard: null
    });
  }

  render() {
    const {places, className} = this.props;
    if (className === `cities__places-list`) {
      return (
        <div className={`${className} places__list tabs__content`}>
          {
            places.map((place) => (
              <CitiesPlaceCard
                key={place.id}
                place={place}
                onMouseEnter={this.handleCardHover}
                onMouseLeave={this.handleCardUnhover}
              />
            ))
          }
        </div>
      );
    } else if (className === `near-places__list`) {
      return (
        <div className={`${className} places__list`}>
          {
            places.map((place) => (
              <NearPlaceCard
                key={place.id}
                place={place}
                onMouseEnter={this.handleCardHover}
                onMouseLeave={this.handleCardUnhover}
              />
            ))
          }
        </div>
      );
    } else {
      return null;
    }
  }
}

PlaceCardList.propTypes = {
  places: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
        propertyName: PropTypes.string.isRequired,
        propertyType: PropTypes.oneOf([`apartment`, `room`, `house`, `hotel`]),
        propertyText: PropTypes.arrayOf(PropTypes.string),
        price: PropTypes.number.isRequired,
        rating: PropTypes.number.isRequired,
        isPremium: PropTypes.bool.isRequired,
        bedroomsAmount: PropTypes.number.isRequired,
        guestMax: PropTypes.number.isRequired,
        propertyItems: PropTypes.arrayOf(PropTypes.string),
        host: PropTypes.shape({
          name: PropTypes.string,
          avatar: PropTypes.string,
          isSuper: PropTypes.bool
        }).isRequired,
        titlePhoto: PropTypes.string.isRequired,
        photos: PropTypes.arrayOf(
            PropTypes.shape({
              title: PropTypes.string,
              id: PropTypes.string
            })
        ),
        reviews: PropTypes.arrayOf(
            PropTypes.shape({
              id: PropTypes.string,
              userName: PropTypes.string,
              avatar: PropTypes.string,
              rating: PropTypes.number,
              text: PropTypes.string
            })
        ),
        coords: PropTypes.arrayOf(PropTypes.number).isRequired
      })
  ),
  className: PropTypes.string.isRequired
};

export default PlaceCardList;

