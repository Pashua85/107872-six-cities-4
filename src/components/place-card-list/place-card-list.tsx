import React from 'react';
import CitiesPlaceCard from '../cities-place-card/cities-place-card';
import NearPlaceCard from '../near-place-card/near-place-card';
import ActionCreator from '../../store/action-creator/action-creator';
import {getActiveOffer} from '../../store/reducers/activeOfferReducer/selectors';
import {connect} from 'react-redux';
import {IPlace} from '../../types/place';

interface PlaceCardListProps {
  places: IPlace[],
  className: string,
  onMouseEnter: (place: IPlace) => void,
  onMouseLeave: () => void,
  activeOffer: null | IPlace,
  onPageLeave: () => void
}

class PlaceCardList extends React.PureComponent<PlaceCardListProps> {
  componentWillUnmount() {
    this.props.onPageLeave();
  }

  render() {
    const {places, className, onMouseEnter, onMouseLeave, activeOffer} = this.props;
    if (className === `cities__places-list`) {
      return (
        <div className={`${className} places__list tabs__content`}>
          {
            places.map((place) => {
              if (place === activeOffer) {
                return (
                  <CitiesPlaceCard
                    key={place.id}
                    place={place}
                    onCardHover={onMouseEnter}
                    onCardUnhover={onMouseLeave}
                    styleObject={{opacity: `.6`}}
                  />
                );
              } else {
                return (
                  <CitiesPlaceCard
                    key={place.id}
                    place={place}
                    onCardHover={onMouseEnter}
                    onCardUnhover={onMouseLeave}
                    styleObject={{}}
                    onCardClick={onCardClick}
                  />
                );
              }
            })
          }
        </div>
      );
    } else if (className === `near-places__list`) {
      return (
        <div className={`${className} places__list`}>
          {
            places.map((place) => {
              if (place === activeOffer) {
                return (
                  <NearPlaceCard
                    key={place.id}
                    place={place}
                    onCardHover={() => {}}
                    onCardUnhover={() => {}}
                    styleObject={{opacity: `.6`}}
                    onCardClick={onCardClick}
                  />
                );
              } else {
                return (
                  <NearPlaceCard
                    key={place.id}
                    place={place}
                    onCardHover={() => {}}
                    onCardUnhover={() => {}}
                    styleObject={{}}
                    onCardClick={onCardClick}
                  />
                );
              }
            })
          }
        </div>
      );
    } else {
      return null;
    }
  }
}

const mapStateToProps = (state) => ({
  activeOffer: getActiveOffer(state)
});

const mapDispatchToProps = (dispatch) => ({
  onMouseEnter: (place) => {
    dispatch(ActionCreator.setActiveOffer(place));
  },
  onMouseLeave: () => {
    dispatch(ActionCreator.setActiveOffer(null));
  },
  onPageLeave: () => {
    dispatch(ActionCreator.setActiveOffer(null));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(PlaceCardList);
export {PlaceCardList};

