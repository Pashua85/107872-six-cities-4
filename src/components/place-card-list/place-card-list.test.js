import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {PlaceCardList} from './place-card-list';
import places from '../../mock/test-offers';

Enzyme.configure({
  adapter: new Adapter()
});

describe(`PlaceCardList`, () => {
  it(`should render correctly for cities pages`, () => {
    const wrapper = shallow(
        <PlaceCardList
          places={places}
          className="cities__places-list"
          onMouseEnter={() => {}}
          onMouseLeave={() => {}}
          onPageLeave={() => {}}
        />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it(`should render correctly for details pages`, () => {
    const wrapper = shallow(
        <PlaceCardList
          places={places}
          className="near-places__list"
          onMouseEnter={() => {}}
          onMouseLeave={() => {}}
          onPageLeave={() => {}}
        />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
