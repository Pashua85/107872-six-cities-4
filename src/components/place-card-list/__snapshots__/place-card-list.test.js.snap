// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaceCardList should render correctly for cities pages 1`] = `ShallowWrapper {}`;
